packages/permission-controller/src/PermissionController.ts

```diff
@@ -1871,15 +1936,45 @@ export class PermissionController<
     const { permissions: approvedPermissions, ...requestData } =
       await this.requestUserApproval(permissionsRequest);
 
-    return [
-      this.grantPermissions({
+    const { permissions: grantedPermissions, inversePatches } =
+      this._grantPermissions({
         subject,
         approvedPermissions,
         preserveExistingPermissions,
         requestData,
-      }),
-      metadata,
-    ];
+      });
+
+    const sideEffects = Object.keys(grantedPermissions).reduce<
```

Comment:
I hadn't thought of the fact that multiple of the granted permissions could have side-effects. Perhaps we should disallow that?

Ideally we would only have one side-effect for the call, because otherwise how do we revert the effects of the first side-effect if the second fails?